_projects/coala_language_server.md

```diff
@@ -42,48 +42,110 @@ a rapidly growing list of editor and language communities:
 We have a state-of-art implementation in
 [coala-vs-code](https://github.com/coala/coala-vs-code) but it is designed
 for VS Code only. We need to design and implement a general mechanism for
-multiple editors/IDEs support.
+multiple editors/IDEs support. We also need clients written to support coala
+langserver for most if not all the supported editors. This is not because of
+the server's inability, but rather the need to configure the front end of the
+editor plugin such as defining how to start the server, communicate with it etc.
 
-The project is divided into two parts:
+The project is divided into three parts:
 
-1. Offer a basic languager server.
+1. Offer a basic language server.
 2. Implement a plugin based on the language server.
+3. Optimize the coala entry point and language server.
 
 #### Features
 
 - Support linting for multiple programming languages.
-- Support 'textDocument/publishDiagnostics', 'textDocument/didChange' and 'textDocument/didSave' requests.
+- The current language server has implementations for:
+  - 'textDocument/didSave'
+  - 'textDocument/didChange'
+  - 'textDocument/publishDiagnostics'
+  and this project will improve and extend the support to at least
+  the following request types:
+  - 'textDocument/didOpen'
+  - 'textDocument/didClose'
+  - 'workspace/formatting'
+  - 'window/showMessageRequest'
 - (Advanced) Multiple plugins Support (e.g. Eclipse Che).
+- Some additional request types that are either out of the scope
+  of this project or are marginally redundant to build but which would
+  improve the user friendliness or functionality  of this language
+  server are:
+  - 'textDocument/willSaveWaitUntil'
+  - 'workspace/didChangeConfiguration'
+  - 'workspace/configuration'
+  - 'workspace/applyEdit'
+  - 'window/showMessage'
+
+#### Interface Optimizations
+
+The current implementation suffers from massive performance
+problems. This is primarily due to the mode in which lang
+server interacts with coala. Currently it works by executing
+coala via cli and redirecting stdout to language server. This
+is not an efficient mechanism because of all the involved overhead.
+This projects aims to provide a practical language server with
+performance improvements. Some ideas currently under consideration
+to achieve the said improvements are:
+
+  - Runtime coala configurations
+  - Language Server specific parameters in coafiles
+  - Custom entry point into coala
```

Comment:
It's not clear how the first two would help address the performance concerns above. A custom entry point into Coala is also a little vague. Could we be explicit about "running Coala in-process"

Reply:
Some bears in coala are long running and some run on entire projects (GlobalBears or ProjectBears). By filtering the coala configuration at runtime and only allowing the certain sections of the configuration to run which are logical to the request type performance can be massively improved. The first two ideas can help with this by allowing a global blacklist for bears not to run in response to some particular request type and by allowing custom configuration of the bears per request type basis in the coafile.

A custom entry point into coala refers to the idea of having a custom pipeline into coala. This way we could prevent the overhead by preprocessing the configuration files, gathering the bears, checking for dependencies and then initializing them. The entire setup can then be reused for subsequent calls.